src/shogun/mathematics/graph/Graph.cpp

```diff
@@ -0,0 +1,185 @@
+#include <shogun/mathematics/graph/Graph.h>
+#include <shogun/mathematics/graph/Tensor.h>
+#include <shogun/mathematics/graph/ops/Input.h>
+
+#include <unordered_set>
+
+using namespace shogun;
+
+Graph::Graph(
+    const std::vector<std::shared_ptr<Input>>& inputs,
+    const std::vector<std::shared_ptr<Node>>& outputs)
+{
+	m_inputs = inputs;
+	m_outputs = outputs;
+}
+
+void Graph::build()
+{
+	auto unordered_nodes = check_fully_connected(m_inputs, m_outputs);
+	build_backend_graph(unordered_nodes);
+}
+
+std::map<std::shared_ptr<Node>, Graph::STATUS> Graph::check_fully_connected(
+    const std::vector<std::shared_ptr<Input>>& inputs,
+    const std::vector<std::shared_ptr<Node>>& outputs)
+{
+	std::deque<std::shared_ptr<Node>> nodes_to_check;
+	for (const auto& node : outputs)
+	{
+		nodes_to_check.push_back(node);
+	}
+	std::unordered_set<std::shared_ptr<Node>> inputs_found;
+	std::map<std::shared_ptr<Node>, STATUS> unordered_nodes;
+
+	while (!nodes_to_check.empty())
+	{
+		auto top_node = nodes_to_check.front();
+		nodes_to_check.pop_front();
+
+		const auto& node_inputs = top_node->get_input_nodes();
+
+		if (!node_inputs.empty())
+		{
+			for (const auto& node : node_inputs)
+				nodes_to_check.push_back(node);
+		}
+		else if (
+		    std::find(inputs.begin(), inputs.end(), top_node) != inputs.end())
+		{
+			// valid input node
+			inputs_found.insert(top_node);
+		}
+		else
+		{
+			// has no children and is not an expected input
+			error("Graph is disconnected in node {}.", *top_node);
+		}
+		unordered_nodes.emplace(top_node, STATUS::UNMARKED);
+	}
+
+	if (inputs_found.size() != inputs.size())
+	{
+		error("Graph found more input tensors than provided.");
+	}
+
+	return unordered_nodes;
+}
+
+void Graph::evaluate(const std::vector<std::shared_ptr<Tensor>>& tensors)
+{
+	auto* env = ShogunEnv::instance();
+	switch (env->graph_backend())
+	{
+	case GRAPH::SHOGUN:
+	{
+		execute_shogun();
+	}
+	break;
+	case GRAPH::NGRAPH:
+	{
+#ifdef USE_NGRAPH
```

Comment:
i guess later we'll want to have this actually be runtime, so that you dont need ngraph to be present during compilation